### PR TITLE
Add support to audio and fluentspeech datasets.

### PR DESCRIPTION
**Pull request by arthurdouillard**
From a discussion with Umberto Cappellazzo.

### CODE REVIEW

**Comment by umbertocappellazzo** [comment]
Hi Arthur, I tried to create an instance of the FluentSpeech class and it works. However, as I iterate through the tasks of the ClassIncremental scenario, I got an AttributeError in the "audio_task_set.py" file. First of all, torch.Long doesn't exist, so I suggest that you change it to torch.FloatTensor(). After having fixed this problem, I got the AttributeError. Below I attach the resulting Traceback: 

File "/Users/umbertocappellazzo/Desktop/PHD/CL_course/untitled1.py", line 27, in <module>
    for x, y, t in train_loader:

  File "/Users/umbertocappellazzo/opt/anaconda3/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 521, in __next__
    data = self._next_data()

  File "/Users/umbertocappellazzo/opt/anaconda3/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 561, in _next_data
    data = self._dataset_fetcher.fetch(index)  # may raise StopIteration

  File "/Users/umbertocappellazzo/opt/anaconda3/lib/python3.9/site-packages/torch/utils/data/_utils/fetch.py", line 49, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]

  File "/Users/umbertocappellazzo/opt/anaconda3/lib/python3.9/site-packages/torch/utils/data/_utils/fetch.py", line 49, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]

  File "/Users/umbertocappellazzo/opt/anaconda3/lib/python3.9/site-packages/continuum/tasks/audio_task_set.py", line 64, in __getitem__
    x = self.get_sample(index)

  File "/Users/umbertocappellazzo/opt/anaconda3/lib/python3.9/site-packages/continuum/tasks/audio_task_set.py", line 60, in get_sample
    return torch.FloatTensor(soundfile.read(self.x[index]))

  File "/Users/umbertocappellazzo/opt/anaconda3/lib/python3.9/site-packages/torch/utils/data/dataset.py", line 83, in __getattr__
    raise AttributeError

**Comment by umbertocappellazzo** [comment]
Hi Arthur, I tried to create an instance of the FluentSpeech class and it works. However, as I iterate through the tasks of the ClassIncremental scenario, I got an AttributeError in the "audio_task_set.py" file. First of all, torch.Long doesn't exist, so I suggest that you change it to torch.FloatTensor(). After having fixed this problem, I got the AttributeError. Below I attach the resulting Traceback: 

File "/Users/umbertocappellazzo/Desktop/PHD/CL_course/untitled1.py", line 27, in <module>
    for x, y, t in train_loader:

  File "/Users/umbertocappellazzo/opt/anaconda3/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 521, in __next__
    data = self._next_data()

  File "/Users/umbertocappellazzo/opt/anaconda3/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 561, in _next_data
    data = self._dataset_fetcher.fetch(index)  # may raise StopIteration

  File "/Users/umbertocappellazzo/opt/anaconda3/lib/python3.9/site-packages/torch/utils/data/_utils/fetch.py", line 49, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]

  File "/Users/umbertocappellazzo/opt/anaconda3/lib/python3.9/site-packages/torch/utils/data/_utils/fetch.py", line 49, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]

  File "/Users/umbertocappellazzo/opt/anaconda3/lib/python3.9/site-packages/continuum/tasks/audio_task_set.py", line 64, in __getitem__
    x = self.get_sample(index)

  File "/Users/umbertocappellazzo/opt/anaconda3/lib/python3.9/site-packages/continuum/tasks/audio_task_set.py", line 60, in get_sample
    return torch.FloatTensor(soundfile.read(self.x[index]))

  File "/Users/umbertocappellazzo/opt/anaconda3/lib/python3.9/site-packages/torch/utils/data/dataset.py", line 83, in __getattr__
    raise AttributeError

**Comment by umbertocappellazzo** [comment]
I think you forgot to include the underscore before x in the line 60 of audio_task_set.py file: return torch.FloatTensor(soundfile.read(self._x[index])). After I fixed it, I got another error in the line 60: **TypeError: not a sequence**.

**Comment by umbertocappellazzo** [comment]
Btw, one possible solution would be letting the user decide whether to return the subintents or not. So we could add a bool variable in the constructor, set to False by default, so that the class returns the subintents together with the intents (a tuple). Lemme know if it's ok for you, else I can just change my personal code to handle this situation.

**Comment by arthurdouillard** [comment]
Hum... If you want multiple outputs, maybe we can use the simpler ContinualScenario, and set as task id one of the possible targets. I'll draft something this afternoon.

**Comment by arthurdouillard** [comment]
I have an idea about that

**Comment by umbertocappellazzo** [comment]
Well, if returning a tuple is a faff, we can just return either the intents or the sub-intents, depending on the need of the user. Usually I expect that someone uses one or the other depending on the situation. You just specify it with a boolean during the initialization. This should do the trick. What do you make of this solution?

**Comment by umbertocappellazzo** [comment]
Yes, it works. One final consideration: what if I'd like to retrieve the transcription associated to a specific label? Is there a chance to get back to them? I need this because I have to carefully design which classes/intents go to a specific experience in the FSC dataset (it's not more complicated than standard images classes). In fact, among the 31 unique intents, some are of the form: "change language to English", "change language to Korean" and two additional ones with Chinese and German. It's pointless including all of them in the first experience for instance, but it would be better to distribute them among different tasks/experiences. I dipped into your code but I think it's not possible. Apart from this detail, it's great!! ✌️

**Comment by umbertocappellazzo** [comment]
Hmm, you are right. As I thought, we could add an attribute to the class and infer the unique intent and the 3 subintents ourself. Or else, having 4 sets (lists), each of them including the values for the intent and the 3 subintents, respectively. Smth like: intent_list = [["change language","Korean", "None"], etc...], action_list = ["activate","increase", etc...], and so on. In this way, if your target tensor is smth like: [13,0,3,4], you just select intent_list[13], action_list[0], and the rest. But this is what came to my mind, don't know if it is easy to implement. You have more experience than me, so do as you think is better, adding the transcriptions as class attribute only should be enough :) Thx again

**Comment by arthurdouillard** [comment]
@TLESORT can you review this PR pls?

**Comment by umbertocappellazzo** [comment]
Ok. Yeah, I meant that one could retrieve the transcription associated to a specific utterance, but it's ok having a class attribute with all the transcription in case someone needs it. Anyhow, I've just realized that if someone wants to retrieve the associated intent or subintents of a sample, (s)he only has to find it in the corresponding attributes of the dataset. For instance, if the label for the intent is "4", you can get back to the original intent by typing `list(dataset.class_ids.keys())[4]` cuz class_ids is a defaultdict. The same applies for the subintents. So this solves my doubt.

**Comment by TLESORT** [comment]
@arthurdouillard can you add some tests, to check if the new taskset is compatible with the other tool of continuum.
Like, slicing, ContinualScenarios, scenario_encoder...
Maybe also adding something about this new taskset in the documentation with its specification / limitation would be nice.

**Comment by umbertocappellazzo** [comment]
Hi Arthur, I just noticed a big issue with the implementation of the FSC dataset (fluentspeech.py). The class doesn't sort the task_ids, so there is a mismatch between the class_ids of the train and test_dataset. In the first commit you made, without the collection.defaultdict, you applied `unique_labels = sorted(list((set([concat_labels[i] for i in range(len(concat_labels))]))))`. Then you forgot to apply something similar to the defaultdicts. In fact, I was testing a model of the first task of a class-incremental scenario (4 classes included), and my results were very bad in the test set. If I print the task_ids of the train dataset and test_dataset created with FluentSpeech(), I get for the train and test datasets:
`defaultdict(<method-wrapper '__next__' of itertools.count object at 0x7fe3d92b54c0>,
            {'change languagenonenone': 0,
             'activatemusicnone': 1,
             'activatelightsnone': 2,
             'deactivatelightsnone': 3,
             'increasevolumenone': 4,
             'decreasevolumenone': 5,
             'increaseheatnone': 6,
             'decreaseheatnone': 7,
             'deactivatemusicnone': 8,
             'activatelampnone': 9,
             'deactivatelampnone': 10,
             'activatelightskitchen': 11,
             'activatelightsbedroom': 12,
             'activatelightswashroom': 13,
             'deactivatelightskitchen': 14,
             'deactivatelightsbedroom': 15,
             'deactivatelightswashroom': 16,
             'increaseheatkitchen': 17,
             'increaseheatbedroom': 18,
             'increaseheatwashroom': 19,
             'decreaseheatkitchen': 20,
             'decreaseheatbedroom': 21,
             'decreaseheatwashroom': 22,
             'bringnewspapernone': 23,
             'bringjuicenone': 24,
             'bringsocksnone': 25,
             'change languageChinesenone': 26,
             'change languageKoreannone': 27,
             'change languageEnglishnone': 28,
             'change languageGermannone': 29,
             'bringshoesnone': 30})`

`defaultdict(<method-wrapper '__next__' of itertools.count object at 0x7fe3c9b93180>,
            {'activatelightsnone': 0,
             'deactivatelightsnone': 1,
             'change languagenonenone': 2,
             'deactivatemusicnone': 3,
             'activatemusicnone': 4,
             'decreasevolumenone': 5,
             'increasevolumenone': 6,
             'increaseheatnone': 7,
             'decreaseheatnone': 8,
             'activatelampnone': 9,
             'deactivatelampnone': 10,
             'activatelightskitchen': 11,
             'activatelightsbedroom': 12,
             'activatelightswashroom': 13,
             'deactivatelightskitchen': 14,
             'deactivatelightsbedroom': 15,
             'deactivatelightswashroom': 16,
             'increaseheatkitchen': 17,
             'increaseheatbedroom': 18,
             'increaseheatwashroom': 19,
             'decreaseheatkitchen': 20,
             'decreaseheatbedroom': 21,
             'decreaseheatwashroom': 22,
             'bringnewspapernone': 23,
             'bringjuicenone': 24,
             'bringsocksnone': 25,
             'bringshoesnone': 26,
             'change languageChinesenone': 27,
             'change languageGermannone': 28,
             'change languageKoreannone': 29,
             'change languageEnglishnone': 30})`

18 intents out of 31 match the others. This means that the corresponding scenarios built upon the two datasets will have wrong classes. Please, check it and tell me if it's the case. Thx

**Comment by arthurdouillard** [comment]
Oh, indeed. I'll try to do that tomorrow, and finish the tests also.

**Comment by arthurdouillard** [comment]
@umbertocappellazzo It's fixed, I've hardcoded the class ids (and action/obj/loc) in the code directly to uniformize across train and test splits.

@TLESORT I've added tests and docs.

File: continuum/scenarios/class_incremental.py [hunk 1]
```diff
@@ -47,11 +47,17 @@ def __init__(
 
     def _setup(self, nb_tasks: int) -> int:
         x, y, _ = self.cl_dataset.get_data()
-        unique_classes = np.unique(y)
+
+        if len(y.shape) > 1:
```
Comment (TLESORT) [inline continuum/scenarios/class_incremental.py:51]: should we verify that task_type is AUDIO, by security?

**Comment by TLESORT** [comment]
LGTM
Good job @arthurdouillard ! Thanks @umbertocappellazzo for helping us by reporting bugs/errors :D

**Pull request merged**